src/IO/Resource/tests/gtest_throttler_constraint.cpp

```diff
@@ -0,0 +1,145 @@
+
+    t.process(start + std::chrono::seconds(1));
+    t.consumed("A", 10);
+
+    t.process(start + std::chrono::seconds(2));
+    t.consumed("A", 10);
+
+    t.process(start + std::chrono::seconds(3));
+    t.consumed("A", 10);
+
+    t.process(start + std::chrono::seconds(4));
+    t.consumed("A", 10);
+
+    t.process(start + std::chrono::seconds(100500));
+    t.consumed("A", 10);
+}
+
+TEST(IOResourceThrottlerConstraint, BucketFilling)
+{
+    ResourceTest t;
+    EventQueue::TimePoint start = std::chrono::system_clock::now();
+    t.process(start, 0);
+
+    t.add<ThrottlerConstraint>("/", "<max_burst>100.0</max_burst><max_speed>10.0</max_speed>");
+    t.add<FifoQueue>("/A", "");
+
+    t.enqueue("/A", {100});
+
+    t.process(start + std::chrono::seconds(0));
+    t.consumed("A", 100); // consume all tokens, but it is still active (not negative)
+
+    t.process(start + std::chrono::seconds(5));
+    t.consumed("A", 0); // There was nothing to consume
+
+    t.enqueue("/A", {10, 10, 10, 10, 10, 10, 10, 10, 10, 10});
+    t.process(start + std::chrono::seconds(5));
+    t.consumed("A", 60); // 5 sec * 10 tokens/sec = 50 tokens + 1 extra request to go below zero
+
+    t.process(start + std::chrono::seconds(100));
+    t.consumed("A", 40); // Consume rest
+
+    t.process(start + std::chrono::seconds(200));
+
+    t.enqueue("/A", {95, 1, 1, 1, 1, 1, 1, 1, 1, 1});
+    t.process(start + std::chrono::seconds(200));
+    t.consumed("A", 101); // check we cannot consume more than max_burst + 1 request
+
+    t.process(start + std::chrono::seconds(100500));
+    t.consumed("A", 3);
+}
+
+TEST(IOResourceThrottlerConstraint, PeekAndAvgLimits)
+{
+    ResourceTest t;
+    EventQueue::TimePoint start = std::chrono::system_clock::now();
+    t.process(start, 0);
+
+    // Burst = 100 token
+    // Peek speed = 50 token/s for 10 seconds
+    // Avg speed = 10 tokens/s afterwards
+    t.add<ThrottlerConstraint>("/", "<max_burst>100.0</max_burst><max_speed>50.0</max_speed>");
+    t.add<ThrottlerConstraint>("/avg", "<max_burst>5000.0</max_burst><max_speed>10.0</max_speed>");
+    t.add<FifoQueue>("/avg/A", "");
+
+    ResourceCost req_cost = 1;
+    ResourceCost total_cost = 10000;
+    for (int i = 0; i < total_cost / req_cost; i++)
+        t.enqueue("/avg/A", {req_cost});
+
+    double consumed = 0;
+    for (int seconds = 0; seconds < 100; seconds++)
+    {
+        t.process(start + std::chrono::seconds(seconds));
+        double arrival_curve = std::min(100.0 + 50.0 * seconds, 5000.0 + 10.0 * seconds) + req_cost;
+        t.consumed("A", static_cast<ResourceCost>(arrival_curve - consumed));
+        consumed = arrival_curve;
+    }
+}
+
+TEST(IOResourceThrottlerConstraint, ThrottlerAndFairness)
+{
+    ResourceTest t;
+    EventQueue::TimePoint start;
+    start += EventQueue::Duration(1000000000);
+    t.process(start, 0);
+
+    t.add<ThrottlerConstraint>("/", "<max_burst>100.0</max_burst><max_speed>10.0</max_speed>");
```

Comment:
Should we also add small case for some extreme setting values, e.g. zeros ?

Reply:
Good point

Reply:
Added